source/common/common/compiled_string_map.md

```diff
@@ -0,0 +1,144 @@
+  branch node at position 12, index 0 = n        ▼
+                                   ┌──┬──┬──┬──┬──┐
+                                   │n │o │p │q │r │
+                                   └┬─┴──┴──┴──┴┬─┘
+                                    │           │
+                                    ▼           ▼
+                       x-envoy-banana         x-envoy-banara
+```
+
+## The lookup
+
+A lookup operation simply walks the generated tree in much the same way as a regular trie.
+For example, given the tree generated above, if you were to search for the header `sponge`,
+the length lookup would find nullptr at index 6, and the resulting value would be the null
+value.
+
+If you were to search for `x-envoy-banaka`, the length would find the branch with most of
+the entries, the index 10 branch would take branch `n`, and the index 12 branch would see that
+`k` is lower than the minimum index in the vector, so the result would be the null value.
+
+If you were to search for `y-envoy-banara`, the length would find the branch with most of
+the entries, the index 10 branch would take branch `n`, the index 12 branch would take
+branch `r`, and the leaf node would do a final string compare, see that
+`y-envoy-barana` != `x-envoy-banara`, and the result would be the null value.
+
+## Performance
+
+In most cases this will be faster than a regular trie, especially for hits - for example,
+instead of 14 steps to match `x-envoy-banana` in the example above, it takes 3 and a
+final memory comparison (which is much faster than one character at a time). For misses,
+some cases will be faster (when the target shares a prefix with an entry), but some will
+be very slightly slower due to more expensive comparisons and dynamic function selection
+in the compiled version.
+
+Benchmark versus a fixed-size 256-branch trie, for the static header map - hit benchmark
+searches for the full range of static headers, miss benchmark used a small set of arbitrary
+non-matching headers:
+
+Benchmark | trie | compiled | trieStdDev | compiledStdDev | change
+-- | -- | -- | -- | -- | --
+bmHeaderMapImplRequestStaticLookupHits | 47.2ns | 16.4ns | 0.629 | 0.378 | -65.3%
+bmHeaderMapImplResponseStaticLookupHits | 34.7ns | 14.3ns | 0.571 | 0.085 | -58.8%
+bmHeaderMapImplRequestStaticLookupMisses | 6.89ns | 6.83ns | 0.044 | 0.034 | -0.01%
+bmHeaderMapImplResponseStaticLookupMisses | 6.40ns | 7.31ns | 0.028 | 0.057 | +14.2%
+
+Versus an `absl::flat_hash_map`, hits are slightly faster and misses are significantly
+faster. (Benchmark has been lost to time.)
+
+In terms of memory, unlike a regular trie, this structure must contain the full keys
+in the leaf nodes, which could potentially make it larger due to, e.g. containing
+`x-envoy-` multiple times where a regular trie only has one 'branch' for that.
+However, it also contains fewer nodes due to not having a node for every character,
+and compared to the prior fixed-size-node 256-branch trie, each node is a lot smaller;
```

Comment:
just a minor comment (no AI):
I think the original trie can be converted to using the same "shrinked-range" (only the needed range of characters) technique. So in terms of memory I think this data-structure is more expensive.
That said, from the CPU perf point-of-view, I tend to agree that this data-structure should perform better, although it is not easy for me to say which of the changes in the data-structure (using per-length buckets, splitting not at the prefix vs condensing a trie) provide the most benefit.

Reply:
Doing just shrunk-range was actually the original PR from which this one forked off as a prerequisite; it made performance a bit worse (more so than this version) for the miss case and made little difference for the hits (worse for short strings, better for long strings). I think this version is also still better memory-wise, because every pointer is 64 bits, and this trie has *far* fewer pointers.

Consider the case of even just one entry, `x-envoy-banana` - in this model it's the root length node and a leaf node, so it's a vector with 13 nullptrs and one pointer (because the length node doesn't do the minimizing thing), a length, and one leaf node containing a value and a string.

In a regular "shrunk" trie, this is 14 nodes each containing a default value (a pointer) and a pointer to the next node, and a vector of chars one character long and a length - so that's 48 pointers, which is significantly larger than the "short path" trie. Even if you cut out the common "x-envoy-" prefix entirely and assume that's shared among so many headers that we can amortize it entirely, it's still 18 pointers just for the 'banana' part of the sequence, which is about the same as the "compiled" trie for the whole thing (which also theoretically shares "common prefixes"!)

Splitting at a more optimal branch point is definitely the main performance enhancer for hits. (It's similar to what gperf does.)

Reply:
Yes, you are correct.
I keep forgetting that this is not a compressed-prefix-trie, but one that has per-character branching.

Reply:
Ah yeah, a compressed-prefix trie (assuming this is a name for "a branch is per-character but also matches the longest string common to all its children") would probably be nearly as fast as the "branch at best position" trie, and a similar size, maybe slightly smaller, since it would also only be a few hops to completion. But it would involve more string-compares.